orttraining/orttraining/models/runner/pipeline.h

```diff
@@ -16,78 +16,196 @@ namespace onnxruntime {
-  //  recorded_events[0]: The first (before backward Send)
-  //  recorded event in backward pass.
-  //  recorded_events[1]: The second (after backward Send)
-  //  recorded event in backward pass.
-  std::vector<int> recorded_events;
+  Pass pass;
+
+  // The last upstream action' time in compute-only schedule.
+  int upstream_time{-1};
+  // The last upstream action' stage in compute-only schedule.
+  int upstream_stage{-1};
+
+  // The rank where this Action is executed on.
+  int this_rank{-1};
+  // The remote rank where this Action interact with.
+  int peer_rank{-1};
+
+  // This action's scheduled time in compute-commute schedule.
+  int full_table_time{-1};
+  // This action's scheduled stage in compute-commute schedule.
+  int full_table_stage{-1};
+};
+
+class PipelineSlot {
+ public:
+  void AddSend(const int batch_id, const PipelineTask::Pass pass, const int upstream_time = -1, const int upstream_stage = -1, const int this_rank = -1, const int peer_rank = -1);
+  void AddRecv(const int batch_id, const PipelineTask::Pass pass, const int upstream_time = -1, const int upstream_stage = -1, const int this_rank = -1, const int peer_rank = -1);
+  void AddCompute(const int batch_id, const PipelineTask::Pass pass, const int upstream_time = -1, const int upstream_stage = -1);
+
+  bool IsEmpty() const { return tasks_.empty(); };
+  size_t NumActions() const { return tasks_.size(); }
+  bool HasCompute() const {
+    for (auto& task : tasks_) {
+      if (task.IsCompute())
+        return true;
+    }
+    return false;
+  }
+
+  bool HasRendTo(const int stage) const {
+    for (auto& task : tasks_) {
+      if (task.IsSendTo(stage)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool HasRecvFrom(const int stage) const {
+    for (auto& task : tasks_) {
+      if (task.IsRecvFrom(stage)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  PipelineTask& operator[](int index);
+  const PipelineTask& operator[](int index) const;
+  PipelineTask& GetFrontAction();
+  const PipelineTask& GetFrontAction() const;
+
+  // Print this structure following a fixed-length format.
+  // It assumes there are at most 2 actions per slot.
+  friend std::ostream& operator<<(std::ostream& stream, const PipelineSlot& slot);
+  void SetWaitedEvent(const std::vector<int> event);
+  std::vector<int> GetWaitedEvent() const;
+  void SetRecordedEvent(const std::vector<int> event);
+  std::vector<int> GetRecordedEvent() const;
+
+ private:
+  // Actions which can be executed in parallel in this time slot.
+  std::vector<PipelineTask> tasks_;
+
+  // For MPI PipeDream schedule, it's used to support Wait -> Recv -> Wait -> Compute -> Record -> Send -> Record.
+  // Since Send, Recv, and Compute are stored in the same slot, each slot contains two waited events and two recorded events.
+  //
+  // For NCCL PipDream schedule, it's used to support Wait -> Recv -> Record -> Wait -> Compute -> Record -> Wait -> Send -> Record.
```

Comment:
>Record [](start = 70, length = 6)

Out of curiosity, why NCCl needs an extra Record after each Recv and Send?

Reply:
For NCC, we need more control on the timing of communication components. It means we need to know the `start` and `end` of each of `Send`, `Recv`, `Forward`, `Backward`.

---
In reply to: [456117125](https://github.com/microsoft/onnxruntime/pull/4524#discussion_r456117125) [](ancestors = 456117125)